### PR TITLE
[RFC] complete: disable folding when completing

### LINKED ISSUE #643: folds close with autocomplete
Create a file that looks like this (main.go)

``` go
package main

import (
    "fmt"
)

func main() {
    fm
    fmt.Print("lmao")
}
```
1. `vim -N -u NONE main.go`
2. `:set foldmethod=indent`
3. `7jA`. 
4. `<c-x><c-l>`. 

Bam the fold gets closed.

### PR DESCRIPTION
**Pull request by nhooyr**
This fixes vim/vim#643, credit to Christian Brabandt

Create a file that looks like this (main.go)

``` go
package main

import (
    "fmt"
)

func main() {
    fm
    fmt.Print("lmao")
}
```
1. `vim -N -u NONE main.go`
2. `:set foldmethod=indent`
3. `7jA`. 
4. `<c-x><c-l>`. 

Bam the fold gets closed.

### CODE REVIEW

File: src/nvim/fold.c [hunk 1]
```diff
@@ -762,6 +762,9 @@ void clearFolding(win_T *win)
  */
 void foldUpdate(win_T *wp, linenr_T top, linenr_T bot)
 {
+  if (disable_fold)
+	  return;
```
Comment (justinmk) [inline src/nvim/fold.c]: can this check `compl_busy` instead (which will need to be moved to globals.h in some fashion)?

**Comment by justinmk** [comment]
Needs a test.

File: src/nvim/fold.c (hunk 1 above)
Comment (nhooyr) [inline src/nvim/fold.c]: probably a better idea, I'll add that in later today with a test.

**Comment by nhooyr** [comment]
@justinmk not sure if I did the test right, but it works.

**Comment by justinmk** [comment]
@nhooyr Great! You're "doing it right" if the test fails before the patch but not after. Is that the case?

**Comment by nhooyr** [comment]
@justinmk yup, of course, that is what I meant by works.

File: src/nvim/globals.h [hunk 2]
```diff
@@ -1039,6 +1039,10 @@ EXTERN int fill_vert INIT(= ' ');
 EXTERN int fill_fold INIT(= '-');
 EXTERN int fill_diff INIT(= '-');
 
+// Set when doing something for completion that may call edit() recursively,
+// which is not allowed. Also used to disable folding during completion
+EXTERN int compl_busy INIT(= false);
```
Comment (justinmk) [inline src/nvim/globals.h]: can you move this under [compl_interrupted](https://github.com/nhooyr/neovim/blob/master/src/nvim/globals.h#L215), seems like that is where the compl-related variables live.

Comment (nhooyr) [inline src/nvim/globals.h]: ok

**Comment by borring** [comment]
What a coincidence! I was having some annoying fold issues a week ago, and I wrote some autocommands to fix them. The problem was that whenever I typed an open brace `{` everything under it would be folded. Also, working on the same file in split windows would mess up the folds in the other split.

So I've devised a way to set fdm to manual whenever in insert mode or when the split isn't in focus.

```
function! IsVimInternalBuffer()
	let l:internalbufs = ["quickfix","help","directory","terminal"]
	for i in internalbufs
		if &buftype == i
			return 1
		endif
	endfor
	return 0
endfunction

function! ToggleFoldmethod(w)
	" Don't do anything when in an internal vim buffer
	" terminal, help, quickfix, etc.. lots of error messages.
	if IsVimInternalBuffer() ==# 1
		return
	endif
	let s:ismanual = match(&foldmethod, '\vmanual')
	if s:ismanual <# 0
		let b:defaultfdm = &foldmethod
		" if a:w is 0, don't toggle to manual
		if a:w ==# 0
			return
		endif
		setl foldmethod=manual
		echo "foldmethod=manual"
	else
		execute "setl foldmethod=" . b:defaultfdm
		echo "foldmethod=" . b:defaultfdm
	endif
endfunction

augroup toggle_fold_manual
	au!
	" When When I am entering text in insert mode, the foldmethod should be
	" manual. I Don't want my folds messing up from me typing open braces.
	autocmd InsertEnter * call ToggleFoldmethod(1)
	autocmd InsertLeave * call ToggleFoldmethod(1)
	" Windows I'm not currently working in should be set to manual folding. I
	" don't want the folds to be jumping around in case I'm working on the same
	" file in a split window. That is the most annoying thing.
	autocmd WinEnter * silent call ToggleFoldmethod(0)
	autocmd WinLeave * silent call ToggleFoldmethod(1)
augroup END
```

**Comment by justinmk** [comment]
@borring Does this PR help you? (BTW, when would buftype=directory?)

**Comment by borring** [comment]
> Does this PR help you?

No, it doesn't. This PR only deals with completion. And it works. I tested it with his line-completion example. I have different problems though.

My problem has to do with text being entered via insert-mode effecting folds.  Example: Typing an opening brace `func blah() {` would immediately fold everything after it until a closing brace is typed.

My solution to that problem was just to switch to `foldmethod=manual` whenever entering insert mode, and toggle back when leaving insert-mode.

Another problem is when I work on the same file in a split window. Even with the insert-mode hack, the things I type into one split will influence the folds in the other split. This necessitated another hacky autocommand. All unfocused windows will be set to `foldmethod=manual` until focus is returned. This way, I could keep an eye on different parts of the file without large swaths of code jumping around and disappearing.

> BTW, when would buftype=directory?

That's a good question. I wrote that function a long time ago for something unrelated. It could be that I was too lazy to do proper research at the time.

**Comment by nhooyr** [comment]
I do something similar but mine is much more simple

```vim
function! s:setFoldManual()
  let b:last_fdm = &foldmethod
  setlocal foldmethod=manual
endfunction
function! s:resFoldMethod()
  let &l:foldmethod = b:last_fdm
  silent! normal zv
endfunction
augroup fold
  autocmd!
  autocmd InsertEnter * call s:setFoldManual()
  autocmd InsertLeave * call s:resFoldMethod()
augroup END
```

We should just disable folding in insert mode? It doesn't sound useful anyway. The auto commands aren't really "hacky".

**Comment by borring** [comment]
Very good feedback! I really only prototyped until it worked, then stopped. I also admittedly don't spend much time in VimL and so I'm not entirely sure I know the syntax. I _definitely_ don't know the api all that well either.

As for calling `match()`, I wanted to script a toggle function, so I copied and pasted an already working toggle I had (which needed `match()`) and didn't bother to change it all that much.

I'll have to remember your comment for when I finally get around to optimizing my vimrc.

On a side note: Do you not have problems with folds while editing (the same file) in split windows?

**Comment by nhooyr** [comment]
@borring I barely ever view the same file in two windows, but I think that it mostly comes down to preference, someone may want it to update automatically instead of moving to the window and bam the folds apply. I think we should just leave it as a autocommand added by the user.

**Pull request merged**